src/infrasys/db_migrations.py

```diff
@@ -0,0 +1,200 @@
+import json
+import sqlite3
+from uuid import uuid4
+
+from loguru import logger
+
+from infrasys import (
+    KEY_VALUE_STORE_TABLE,
+    TIME_SERIES_ASSOCIATIONS_TABLE,
+    TIME_SERIES_METADATA_TABLE,
+)
+from infrasys.utils.sqlite import execute
+from infrasys.utils.time_utils import _str_timedelta_to_iso_8601
+
+TEMP_TABLE = "legacy_metadata"
+
+
+def needs_migration(conn: sqlite3.Connection, version: str | None = None) -> bool:
+    """Check if time series metadata table needs migration."""
+    query = "SELECT * FROM sqlite_master"
+    result = conn.execute(query).fetchall()[0]
+    if "time_series_associations" in result:
+        return False
+    return True
+
+
+def migrate_legacy_schema(conn: sqlite3.Connection) -> bool:
+    """Migrate from legacy schema to new schema with separated metadata.
+
+    Notes
+    -----
+    This migration:
+    1. Creates temporary table.
+    2. Extracts features from metadata and saves as features
+    3. Converts resolution to ISO format
+    4. Preserves all data while restructuring the database
+
+    Returns
+    -------
+        bool: True if migration was successful
+    """
+    logger.info("Migrating legacy metadata schema.")
+
+    legacy_columns = [
+        "id",
+        "time_series_uuid",
+        "time_series_type",
+        "initial_time",
+        "resolution",
+        "variable_name",
+        "component_uuid",
+        "component_type",
+        "user_attributes_hash",
+        "metadata",
+    ]
+
+    cursor = conn.cursor()
+    cursor.execute(f"SELECT * FROM {TIME_SERIES_METADATA_TABLE} LIMIT 1")
+    columns = [desc[0] for desc in cursor.description]
+    if not all(column in columns for column in legacy_columns):
+        logger.error(f"Legacy schema does not match expected columns: {columns}")
+        msg = "Bug: Legacy schema doesn't match expected structure"
+        raise NotImplementedError(msg)
+
+    logger.info("Creating backup tables.")
+    execute(
+        cursor,
+        f"CREATE TEMPORARY TABLE {TEMP_TABLE} AS SELECT * FROM {TIME_SERIES_METADATA_TABLE}",
+    )
+    execute(
+        cursor,
+        f"DROP TABLE {TIME_SERIES_METADATA_TABLE}",
+    )
+
+    logger.info("Creating new schema tables...")
+    execute(
+        cursor,
+        f"""
+        CREATE TABLE {TIME_SERIES_METADATA_TABLE} (
+            id INTEGER PRIMARY KEY,
+            metadata_uuid TEXT,
+            metadata JSON NOT NULL
+        )
+    """,
+    )
+    execute(
+        cursor, f"CREATE TABLE {KEY_VALUE_STORE_TABLE}(key TEXT PRIMARY KEY, VALUE JSON NOT NULL)"
+    )
+
+    execute(
+        cursor,
+        f"""
+        CREATE TABLE {TIME_SERIES_ASSOCIATIONS_TABLE} (
+            id INTEGER PRIMARY KEY,
```

Comment:
Should we synchronize this with `TimeSeriesMetadataStore._create_associations_table`? I can imagine someone changing that one but not this one.

Reply:
Good suggestion. I added a function that we will use for both the migration and the metadata store.